Libraries/NativeAnimation/RCTNativeAnimatedModule.m

```diff
@@ -15,7 +15,13 @@
 @implementation RCTNativeAnimatedModule
 {
   RCTNativeAnimatedNodesManager *_nodesManager;
+  // Oparations called after views have been updated.
   NSMutableArray<AnimatedOperation> *_operations;
+  // Operations called before views have been updated.
+  NSMutableArray<AnimatedOperation> *_preOperations;
+  // Lock used for _operations and _preOperations. This is needed since operations
+  // are added from the UIManager queue and executed from the main queue.
+  NSLock *_operationsLock;
```

Comment:
I would prefer to come up with a solution that doesn't use the lock on each operation to further ensure the code is correct, plus perhaps better performance due to less contention.

My concern is that we may process half a JS frame's worth of operations, rendering an inconsistent state to the user, regardless of whether the JS thread runs faster than the UI thread or vice versa. I believe we should treat all the operations in a given JS frame as one atomic set -- the code should be structured in a way that makes it impossible to do otherwise.

Here is one idea: we split `_operations` into two arrays (same for `_preOperations`): `NSMutableArray<AnimOp> *_operations` and `NSMutableArray<NSArray<AnimOp> *> *_operationLists`. The first array is accessible only from the UIManager queue and the second only from the native main queue (mostly).

The UIManager queue still writes to `_operations` the same way. When we know it has written all of its operations for the current JS frame, we run:

```objc
// On the UI queue
[_operationsLock lock];
// Add both operation sets to be accessible from the main queue
[_preOperationsLists addObject[_preOperations];
[_operationsLists addObject:_operations];
// We are on the UIManager queue at the end of the frame, this is safe:
_preOperations = [NSMutableArray new];
_operations = [NSMutableArray new];
[_operationsLock unlock];
```

There's a little trickiness on the main queue to also use the above lock (plus probably more efficient solutions than what I proposed).